File: algorithms/root-finding-optimization/newtons-method/python/newtons_method.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""newtons_method.py

Newton's method is an iterative root-finding algorithm that starts with an initial guess and refines it iteratively to find a root of a given function. Here's how the algorithm works:

    Initialization: Start with an initial guess x0x0​.

    Iteration: Repeat the following steps until convergence or a maximum number of iterations:
        Compute the change in xx using the formula: Δx=−f(x)f′(x)Δx=−f′(x)f(x)​, where f(x)f(x) is the function and f′(x)f′(x) is its derivative.
        Update the value of xx using x:=x+Δxx:=x+Δx.
        Check for convergence: If ∣Δx∣<tol∣Δx∣<tol, where tol is a predefined tolerance, exit the loop.

    Termination: Return the final value of xx and the number of iterations.

In the example usage, we demonstrate how to find a root of a simple function f(x)=x2−4f(x)=x2−4 using Newton's method. We provide the function f(x)f(x), its derivative f′(x)f′(x), an initial guess x0x0​, and call the newtons_method function to obtain the root and the number of iterations required for convergence. Adjustments to the initial guess, tolerance, and maximum number of iterations can be made based on the specific problem being solved.

.. _PEP 0000:
    https://peps.python.org/pep-0000/
"""


from typing import Callable, Tuple
import numpy as np


def main() -> None:
    """Driving code and main function"""
    # Example: Find a root of a function using Newton's method
    def f(x: float) -> float:
        return x**2 - 4

    def df(x: float) -> float:
        return 2 * x

    x0 = 2.0
    root, iterations = newtons_method(f, df, x0)
    print("Root:", root)
    print("Number of iterations:", iterations)
 
    return None


def newtons_method(f: Callable[[float], float], df: Callable[[float], float], x0: float, tol: float = 1e-6, max_iter: int = 1000) -> Tuple[float, int]:
    """Find a root of a function using Newton's method.

    Args:
        f (Callable[[float], float]): The function for which to find the root.
        df (Callable[[float], float]): The derivative of the function.
        x0 (float): Initial guess for the root.
        tol (float, optional): Tolerance for convergence. Defaults to 1e-6.
        max_iter (int, optional): Maximum number of iterations. Defaults to 1000.

    Returns:
        Tuple[float, int]: The root and the number of iterations.
    """
    x = x0
    for i in range(max_iter):
        delta_x = -f(x) / df(x)
        x += delta_x
        if abs(delta_x) < tol:
            break
    return x, i + 1


if __name__ == '__main__':
    main()

File: algorithms/root-finding-optimization/newtons-method/java/NewtonsMethod.java
import java.util.function.DoubleUnaryOperator;

public class NewtonsMethod {

    /**
     * Find a root of a function using Newton's method.
     *
     * @param f      The function for which to find the root.
     * @param df     The derivative of the function.
     * @param x0     Initial guess for the root.
     * @param tol    Tolerance for convergence.
     * @param maxItr Maximum number of iterations.
     * @return The root and the number of iterations.
     */
    public static Result newtonsMethod(DoubleUnaryOperator f, DoubleUnaryOperator df, double x0, double tol, int maxItr) {
        double x = x0;
        double delta_x;
        for (int i = 0; i < maxItr; i++) {
            delta_x = -f.applyAsDouble(x) / df.applyAsDouble(x);
            x += delta_x;
            if (Math.abs(delta_x) < tol) {
                return new Result(x, i + 1);
            }
        }
        return new Result(x, maxItr);
    }

    public static void main(String[] args) {
        // Example: Find a root of a function using Newton's method
        DoubleUnaryOperator f = x -> x * x - 4;
        DoubleUnaryOperator df = x -> 2 * x;
        double x0 = 2.0;
        double tol = 1e-6;
        int maxItr = 1000;

        Result result = newtonsMethod(f, df, x0, tol, maxItr);
        System.out.println("Root: " + result.root);
        System.out.println("Number of iterations: " + result.iterations);
    }

    static class Result {
        double root;
        int iterations;

        Result(double root, int iterations) {
            this.root = root;
            this.iterations = iterations;
        }
    }
}
